### PR TITLE
Request to review seqtk module

### PR DESCRIPTION
**Pull request by kaurravneet4123**
- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, include test data in your PR.
- [ ] Remove all TODO statements.
- [ ] Emit the `<SOFTWARE>.version.txt` file.
- [ ] Follow the naming conventions.
- [ ] Follow the parameters requirements.
- [ ] Follow the input/output options guidelines.
- [ ] Add a resource `label`
- [ ] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
    - [ ] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [ ] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [ ] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd`

### CODE REVIEW

**Comment by kaurravneet4123** [comment]
Hi @drpatelh, I have made a pull request for seqtk module. I cannot see the results of conda/docker/singularity in the pull request. If anyone can kindly have a look, that would be great. Many thanks, Rav.

**Comment by drpatelh** [comment]
Hi @Rav! This module should be called `seqtk/sample` and not just `seqtk` because it has multiple sub-commands. I will close this for now so you can submit a fresh PR. It will be easier to create from scratch with `nf-core modules create`. WIll take a look at the rest of the code and comment here though.

File: software/seqtk/main.nf [hunk 1]
```diff
@@ -0,0 +1,48 @@
+// Import generic module functions
+include { initOptions; saveFiles; getSoftwareName } from './functions'
+
+params.options = [:]
+options        = initOptions(params.options)
+
+def VERSION = '1.3'
+
```
Comment (drpatelh) [inline software/seqtk/main.nf:8]:
```suggestion
```

File: software/seqtk/main.nf [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+process SEQTK {
```
Comment (drpatelh) [inline software/seqtk/main.nf:9]:
```suggestion
process SEQTK_SAMPLE {
```

File: software/seqtk/main.nf [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    conda (params.enable_conda ? "bioconda::seqtk=1.3" : null)
+    if (workflow.containerEngine == 'singularity' && !params.singularity_pull_docker_container) {
+        container "https://depot.galaxyproject.org/singularity/seqtk:1.3--h5bf99c6_3"
+    } else {
+        container "quay.io/biocontainers/seqtk:1.3--h5bf99c6_3"
+    }
+
+    input:
+    tuple val(meta), path(reads)
+
```
Comment (drpatelh) [inline software/seqtk/main.nf:25]:
```suggestion
    tuple val(meta), path(reads)
    val   sample_size
    
```

File: software/seqtk/main.nf [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    output:
+    tuple val(meta), path("*.fastq.gz"), emit: reads
+    path "*.version.txt"          , emit: version
```
Comment (drpatelh) [inline software/seqtk/main.nf:28]:
```suggestion
    tuple val(meta), path("*.fastq.gz"), emit: reads
    path "*.version.txt"               , emit: version
```

File: software/seqtk/main.nf [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+    script:
+    def software = getSoftwareName(task.process)
+    def prefix   = options.suffix ? "${meta.id}${options.suffix}" : "${meta.id}"
+    if (meta.single_end) {
+        """
+        [ ! -f  ${prefix}.fastq.gz ] && ln -s $reads ${prefix}.fastq.gz
+        seqtk sample $options.args ${prefix}.fastq.gz | gzip > ${prefix}.fastq.gz
+        echo $VERSION > ${software}.version.txt
```
Comment (drpatelh) [inline software/seqtk/main.nf:37]:
```suggestion
        seqtk \\
	    sample \\
            $options.args \\
            $reads \\ 
            $sample_size \\
            | gzip > ${prefix}.fastq.gz

	echo \$(seqtk 2>&1) | sed 's/^.*Version: //; s/ .*\$//' > ${software}.version.txt
```

File: tests/software/seqtk/main.nf [hunk 6]
```diff
@@ -0,0 +1,28 @@
+#!/usr/bin/env nextflow
+
+nextflow.enable.dsl = 2
+
+include { SEQTK } from '../../../software/seqtk/main.nf' addParams( options: [ 'args': '-s100 250000' ] )
```
Comment (drpatelh) [inline tests/software/seqtk/main.nf:5]:
```suggestion
include { SEQTK } from '../../../software/seqtk/main.nf' addParams( options: [ 'args': '-s100', 'suffix':'sampled' ] )
```

File: tests/software/seqtk/main.nf [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+/*
+ * Test with single-end data
+ */
+workflow test_seqtk_single_end {
+    input = [ [ id:'test', single_end:true ], // meta map
+              file(params.test_data['sarscov2']['illumina']['test_1_fastq_gz'], checkIfExists: true) ]
+
+    SEQTK ( input )
```
Comment (drpatelh) [inline tests/software/seqtk/main.nf:14]:
```suggestion
    SEQTK ( input,  250000)
```

File: tests/software/seqtk/main.nf [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+/*
+ * Test with paired-end data
+ */
+workflow test_seqtk_paired_end {
+    input = [ [id: 'test', single_end: false], // meta map
+              [ file(params.test_data['sarscov2']['illumina']['test_1_fastq_gz'], checkIfExists: true),
+                file(params.test_data['sarscov2']['illumina']['test_2_fastq_gz'], checkIfExists: true) ]
+            ]
+
+    SEQTK ( input )
```
Comment (drpatelh) [inline tests/software/seqtk/main.nf:26]:
```suggestion
    SEQTK ( input, 250000 )
```

**Review by drpatelh — COMMENTED** [review]
Ok. Added a bunch of comments so hopefully you get the idea now. May have to play with the module a little to get the tests to pass and need to add docs to `meta.yml` but otherwise should work.

**Comment by kaurravneet4123** [comment]
many thanks @drpatelh for your time and suggestions.
I will create a new template accordingly.
Have a nice weekend!
signature   Ravneet 

    On Friday, May 7, 2021, 01:03:52 PM GMT+1, Harshil Patel ***@***.***> wrote:  
 
 


@drpatelh commented on this pull request.

In tests/software/seqtk/test.yml:
> +    - path: output/seqtk/test_1.fastq.gz
+    - path: output/seqtk/test_2.fastq.gz
⬇️ Suggested change-    - path: output/seqtk/test_1.fastq.gz
-    - path: output/seqtk/test_2.fastq.gz
+    - path: output/seqtk/test.sampled_1.fastq.gz
+    - path: output/seqtk/test.sampled_2.fastq.gz

**Pull request closed without merging**